clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -685,6 +705,13 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
                             "to request metadata update now", batch.topicPartition,
                             error.exception(response.errorMessage).toString());
                 }
+                if (error == Errors.NOT_LEADER_OR_FOLLOWER || error == Errors.FENCED_LEADER_EPOCH) {
```

Comment:
It's strange that there's a mixture of checking errors with `instanceof` and `Errors.ARBITRARY_KAFKA_ERROR`. Perhaps this line ought to use `instanceof` too.